bundles/org.eclipse.swt/Eclipse SWT Drag and Drop/cocoa/org/eclipse/swt/dnd/DropTarget.java

```diff
@@ -434,7 +442,15 @@ static long dropTargetProc(long id, long sel, long arg0) {
 	if (display == null || display.isDisposed()) return 0;
 	Widget widget = display.findWidget(id);
 	if (widget == null) return 0;
-	DropTarget dt = (DropTarget)widget.getData(DND.DROP_TARGET_KEY);
+	DropTarget dt = (DropTarget) widget.getData(DND.DROP_TARGET_KEY);
+	if (dt == null) {
+		while (widget != null && (widget instanceof Label || !(widget instanceof Shell))) {
```

Comment:
Given that a Label is not a Shell, it's the `widget instanceof Label` redundant?

Later we assume (via the cast) that widget is a Control.  So wouldn't it be better to test that?

Reply:
`instanceof` check on shell is actually to break out of the loop( when we have finished iterating through the ancestry chain of the widget with a null drop target key) and hence it is required. But on a second thought that check should preceed the `instanceof` check on label :

`while (widget != null && !(widget instanceof Shell) && widget instanceof Label) {`

as the SWT binary repo has been moved I am setting up my Dev env to test this change.

Reply:
I think you are trying to walk up the containers and to stop when you get to a shell.   The proposal you show above can't possibly be correct because it can only be true for a Label so will not go up a parent chain.  This looks questionable.

Reply:
@merks agree, I am trying to walk yo the container chain up to the shell, and pick choose the youngest ancestor that has a non null drop target key . And I want to do this only for label widgets (as per the reported problem).

I amended my logic accordingly. I tested this locally and everything works fine. Please have a look.